### PR TITLE
DM-4573 expression support (converted expression parsing library to JS)

### PR DESCRIPTION
**Pull request by tgoldina**
Converted GWT expression library to JS.
Wrote a unit test for its functions and operators.
Added expression support to XY plot column selection.

### CODE REVIEW

File: src/firefly/js/util/expr/Parser.js [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+ //
+ //<P>Whitespace outside identifiers is ignored.
+ //
+ //<P>Examples:
+ //<UL>
+ //<LI>42
+ //<LI>2-3
+ //<LI>cos(x^2) + sin(x^2)
+ //<UL>
+
```
Comment (loitly) [inline src/firefly/js/util/expr/Parser.js]: shouldn't these comments be in a jsdoc block?

File: src/firefly/js/visualize/XYPlotOptions.jsx [hunk 2]
```diff
@@ -11,12 +11,37 @@ import {FieldGroup} from '../ui/FieldGroup.jsx';
...
+    for (i = text.length-1; i>=0; i--) {
+        c = text.charAt(i);
+        if (!/[A-Za-z\d_]/.test(c)) {
+            priorIdx = i;
+            break;
+        }
+    }
+    if (priorIdx > 0) priorContent = text.substring(0, priorIdx+1);
+    if (priorIdx < text.length) token = text.substring(priorIdx+1);
+
```
Comment (loitly) [inline src/firefly/js/visualize/XYPlotOptions.jsx]: i am sure this works fine.
just sharing info.. have you tried using string.match?  works great for taking chunks out of matching strings..
i.e.  
> 'cos(x^2)'.match(/(.*)([^A-Za-z\d_])/)
[ 'cos(x^2)', 'cos(x^2', ')', index: 0, input: 'cos(x^2)' ]

**Comment by loitly** [comment]
A couple of things I noticed:

1.  It performs badly in Firefly. There is too much lags between keystrokes.  Not sure what you can do about that.
2.  Selecting points in xy-plot will highlight table, but not vise-versa.  Is that intentional?

Otherwise, it is fine.  I will put this ticket into reviewed.  If you made changes and wanted me to review again, feel free to re-assign.

**Comment by tgoldina** [comment]
Thank you for testing. I've fixed highlight from table. (I forgot to add
support for expressions when calculating highlight.)

Also thank you for suggesting [match](
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/match),
it simplifies things.

Regarding Firefox performance, I noticed I still get unexpected table
reloads. (For example, when I enter or remove a key in a suggest box.) When
only plot is on the page (as in expanded view), it seems to perform much
better.


2016-04-28 18:08 GMT-07:00 loitly <notifications@github.com>:

**Pull request merged**